### PR TITLE
chromium: fix ANGLE vulkan backend on NixOS

### PR DESCRIPTION
**Pull request by Kiskae**
## Description of changes

libANGLE includes an unpatched copy of `libvulkan.so.1`, so on NixOS it is unable to locate the runtime icds.
By replacing this copy with a symlink to our own version of `vulkan-loader`, it now correctly loads the system icds.

Output of `VK_LOADER_DEBUG=driver` before change:
```console
...
DRIVER:               Found no files
ERROR | DRIVER:    vkCreateInstance: Found no drivers!
...
DRIVER | LAYER:    vkCreateDevice layer callstack setup to:
DRIVER | LAYER:       <Application>
DRIVER | LAYER:         ||
DRIVER | LAYER:       <Loader>
DRIVER | LAYER:         ||
DRIVER | LAYER:       <Device>
DRIVER | LAYER:           Using "SwiftShader Device (Subzero)" with driver: "/nix/store/byiqfj6ays9jazj4b71akvvgklww3q26-electron-unwrapped-28.2.2/libexec/electron/././libvk_swiftshader.so"
```

Output of `VK_LOADER_DEBUG=driver` after change:
```console
...
DRIVER:               Found the following files:
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/dzn_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/intel_hasvk_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/intel_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/lvp_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/nvidia_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/radeon_icd.x86_64.json
DRIVER:                  /run/opengl-driver/share/vulkan/icd.d/virtio_icd.x86_64.json
...
DRIVER | LAYER:    vkCreateDevice layer callstack setup to:
DRIVER | LAYER:       <Application>
DRIVER | LAYER:         ||
DRIVER | LAYER:       <Loader>
DRIVER | LAYER:         ||
DRIVER | LAYER:       <Device>
DRIVER | LAYER:           Using "NVIDIA GeForce RTX 3060" with driver: "/nix/store/d0j0ppbwzkpyhpi8mrgdcqb4x45sry5g-nvidia-x11-545.29.02-6.6.1/lib/libGLX_nvidia.so.0"
```

The only concern is the fallback behavior if there is no working vulkan driver installed. Chromium bundled its own icd at `$out/libexec/<app>/vk_swiftshader_icd.json`. We might be able to fix this by setting the `VK_ADD_DRIVER_FILES` env variable in the wrapper.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilylange** [comment]
Thanks :)

This certainly looks like it would fix that bug (for which I haven't had time yet to look into).

Will try to look into your PR over the next two weeks or so (chromium does not have enough maintainers to look into this quicker, sorry).

**Comment by emilylange** [comment]
@ofborg eval

**Comment by Kiskae** [comment]
> Thanks :)
> 
> This certainly looks like it would fix that bug (for which I haven't had time yet to look into).
> 
> Will try to look into your PR over the next two weeks or so (chromium does not have enough maintainers to look into this quicker, sorry).

No problem. This PR is the result of me finding the time to figure out what their bundled libvulkan was doing. So it's a theme :)

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (bjornfor) [inline pkgs/applications/networking/browsers/chromium/common.nix:488]: Nit: quote shell variables.

**Comment by Kiskae** [comment]
Quoted shell variables and rebased to a newer version of nixpkgs.

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (emilylange) [inline pkgs/applications/networking/browsers/chromium/common.nix:484]: Is there a reason why you check whether this file exists before applying your fix?

If this file ever changes, we won't be able to catch it just from the build.
And if there are e.g. electron-source versions where this file is missing, we should probably use version conditionals instead.

```suggestion
```

File: pkgs/development/tools/electron/binary/generic.nix
Comment (emilylange) [inline pkgs/development/tools/electron/binary/generic.nix:1]: Can't comment on electron-bin builds.

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (emilylange) [inline pkgs/applications/networking/browsers/chromium/common.nix:1]: And I am also not sure if we what to add more such things to the unwrapped drv in `common.nix`.

It makes iterating on changes more time-consuming, since you literally have to build a full chromium from scratch and only then are able to see if it works.

How about we move this to the mess that calls itself `pkgs/applications/networking/browsers/chromium/default.nix` instead? 🫠

Or maybe I am missing something here.

**Review by emilylange — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (Kiskae) [reply pkgs/applications/networking/browsers/chromium/common.nix:1]: The problem is that the path gets compiled into the consumer of the angle library, so chromium (or libgles, not entirely sure at the moment) directly imports `$libExec/libvulkan` instead of using libdl's resolution algorithm.

So any changes made in the wrapper would have no effect on the loader that is used.

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (Kiskae) [reply pkgs/applications/networking/browsers/chromium/common.nix:484]: I am probably being overly cautious, it is fine to remove the check.

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (Kiskae) [reply pkgs/applications/networking/browsers/chromium/common.nix:1]: https://github.com/NixOS/nixpkgs/issues/272912#issuecomment-1847877119

Here I got an ltrace for electron, but this behavior should be identical for chromium.

Comment (emilylange) [reply pkgs/applications/networking/browsers/chromium/common.nix:1]: Hmm I see. I mean something like this (very hacky, please don't copy, just so you get the idea):

```diff
diff --git a/pkgs/applications/networking/browsers/chromium/default.nix b/pkgs/applications/networking/browsers/chromium/default.nix
index 9da0f725ed56..375df59a5fcf 100644
--- a/pkgs/applications/networking/browsers/chromium/default.nix
+++ b/pkgs/applications/networking/browsers/chromium/default.nix
@@ -1,4 +1,4 @@
-{ newScope, config, stdenv, fetchurl, makeWrapper
+{ newScope, config, stdenv, fetchurl, makeWrapper, vulkan-loader
 , buildPackages
 , ed, gnugrep, coreutils, xdg-utils
 , glib, gtk3, gtk4, gnome, gsettings-desktop-schemas, gn, fetchgit
@@ -161,7 +161,15 @@ let
         chmod u+w $out/libexec/chromium
         cp -a ${widevineCdm}/WidevineCdm $out/libexec/chromium/
       ''
-    else browser;
+    else runCommand (browser.name + "-vulkan") { version = browser.version; } ''
+        mkdir -p $out
+        cp -a ${browser}/* $out/
+        chmod u+w $out/libexec/chromium
+
+        rm $out/libexec/chromium/libvulkan.so.1
+
+        ln -s -t "$out/libexec/chromium" "${lib.getLib vulkan-loader}/lib/libvulkan.so.1"
+      '';
 
 in stdenv.mkDerivation {
   pname = lib.optionalString ungoogled "ungoogled-"
@@ -187,7 +195,7 @@ in stdenv.mkDerivation {
 
   buildCommand = let
     browserBinary = "${chromiumWV}/libexec/chromium/chromium";
-    libPath = lib.makeLibraryPath [ libva pipewire wayland gtk3 gtk4 libkrb5 ];
+    libPath = lib.makeLibraryPath [ libva pipewire wayland gtk3 gtk4 libkrb5 vulkan-loader ];
 
   in with lib; ''
     mkdir -p "$out/bin"
```

Comment (emilylange) [reply pkgs/applications/networking/browsers/chromium/common.nix:1]: But on second thought:
I am fine with your approach. Will merge if you remove the file exists conditional and my nixpkgs-review then reports everything builds :)

File: pkgs/development/tools/electron/binary/generic.nix
Comment (Kiskae) [reply pkgs/development/tools/electron/binary/generic.nix:1]: should I move these changes to a follow-up PR, since it is essentially duplicating the chromium changes to the binary releases of electron?

File: pkgs/applications/networking/browsers/chromium/common.nix
Comment (Kiskae) [reply pkgs/applications/networking/browsers/chromium/common.nix:1]: That might work, I believe libangle uses `dladdr` on the primary executable to resolve the absolute path. But that might be better saved for some future exploration since it might have other unintended consequences.

If it ends up working it might be worth moving all of the ' fixup' phase into the wrapper since that would be where a lot of the experimental nixos fixes would happen.

**Review by emilylange — APPROVED** [review]
Result of `nixpkgs-review pr 293063` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>archivebox</li>
    <li>archivebox.dist</li>
    <li>clockify</li>
    <li>electron-fiddle</li>
    <li>electron_24</li>
    <li>electron_24-bin</li>
    <li>itch</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>authy</li>
    <li>redisinsight</li>
  </ul>
</details>
<details>
  <summary>84 packages built:</summary>
  <ul>
    <li>affine</li>
    <li>aws-azure-login</li>
    <li>bilibili</li>
    <li>bitwarden-desktop</li>
    <li>bitwarden-directory-connector</li>
    <li>blockbench</li>
    <li>breitbandmessung</li>
    <li>bruno</li>
    <li>camunda-modeler</li>
    <li>chromium</li>
    <li>chromium.sandbox</li>
    <li>deltachat-desktop</li>
    <li>drawio</li>
    <li>drawio-headless</li>
    <li>electron (electron_29)</li>
    <li>electron-bin (electron_29-bin)</li>
    <li>electron_27</li>
    <li>electron_27-bin</li>
    <li>electron_28</li>
    <li>electron_28-bin</li>
    <li>element-desktop</li>
    <li>element-desktop-wayland</li>
    <li>fast-cli</li>
    <li>feishin</li>
    <li>freetube</li>
    <li>headset</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>jitsi-meet-electron</li>
    <li>kuro</li>
    <li>logseq</li>
    <li>mattermost-desktop</li>
    <li>mermaid-cli</li>
    <li>mermaid-filter</li>
    <li>micropad</li>
    <li>morgen</li>
    <li>nix-tour</li>
    <li>obsidian</li>
    <li>open-stage-control</li>
    <li>pandoc-drawio-filter</li>
    <li>pandoc-drawio-filter.dist</li>
    <li>passky-desktop</li>
    <li>percollate</li>
    <li>playwright-test</li>
    <li>pocket-casts</li>
    <li>podman-desktop</li>
    <li>pritunl-client</li>
    <li>protonmail-desktop</li>
    <li>puppeteer-cli</li>
    <li>python311Packages.mkdocs-drawio-exporter</li>
    <li>python311Packages.mkdocs-drawio-exporter.dist</li>
    <li>python311Packages.pytest-playwright</li>
    <li>python311Packages.pytest-playwright.dist</li>
    <li>python312Packages.mkdocs-drawio-exporter</li>
    <li>python312Packages.mkdocs-drawio-exporter.dist</li>
    <li>python312Packages.pytest-playwright</li>
    <li>python312Packages.pytest-playwright.dist</li>
    <li>r2modman</li>
    <li>revolt-desktop</li>
    <li>sharedown</li>
    <li>single-file-cli</li>
    <li>sitespeed-io</li>
    <li>standardnotes</li>
    <li>stretchly</li>
    <li>super-productivity</li>
    <li>teams-for-linux</li>
    <li>terra-station</li>
    <li>thedesk</li>
    <li>threema-desktop</li>
    <li>uhk-agent</li>
    <li>uhk-udev-rules</li>
    <li>uivonim</li>
    <li>ungoogled-chromium</li>
    <li>ungoogled-chromium.sandbox</li>
    <li>vesktop</li>
    <li>vhs</li>
    <li>vieb</li>
    <li>wayback</li>
    <li>webcord</li>
    <li>webcord-vencord</li>
    <li>webtorrent_desktop</li>
    <li>whalebird</li>
    <li>youtube-music</li>
    <li>ytdownloader</li>
  </ul>
</details>

~~~bash
# git describe --abbrev=40
24.05-pre-59795-g6bffef7a1a02608183e05b19294068d01af22075
# nix-build nixos/tests/chromium.nix -A stable -A ungoogled
/nix/store/ci595gqcqa0v1wvd3fjggvqlyfc5q06a-vm-test-run-chromium-stable
/nix/store/282lpy974sm153invsa0qnh55dfmzs9a-vm-test-run-chromium-ungoogled
~~~

**Pull request merged**
